File: Katas/src/main/java/solution/NameInverter.java
package solution;

import com.google.common.collect.Lists;

import java.util.ArrayList;
import java.util.List;

public class NameInverter {

  public static String invert(String name) {
    List<String> nameParts = breakingIntoPartsIgnoringWhitespace(name);

    if (nameParts.size() < 2)
      return nameParts.get(0);

    return invert(withoutHonorifics(nameParts));
  }

  private static ArrayList<String> breakingIntoPartsIgnoringWhitespace(String name) {
    return Lists.newArrayList(name.trim().split(RegularExpressions.ANY_SPACES));
  }

  private static List<String> withoutHonorifics(List<String> nameParts) {
    String s = nameParts.get(0);

    if (Honorifics.isHonorific(s))
      nameParts.remove(0);

    return nameParts;
  }

  private static String invert(List<String> nameParts) {
    String first = nameParts.get(0);
    String last = nameParts.get(1);
    String postnominals = "";
    postnominals = findAndMergePostnominals(nameParts);

    return String.format("%s, %s %s", last, first, postnominals).trim();
  }

  private static String findAndMergePostnominals(List<String> nameParts) {
    String postnominals = "";

    for (String each : nameParts.subList(2, nameParts.size()))
      postnominals += each + " ";

    return postnominals;
  }
}

class RegularExpressions {
  public static final String ANY_SPACES = "\\s+";
}

class Honorifics {
  private static final List<String> KNOWN_HONORIFICS = Lists.newArrayList("Mr.", "Mrs.");

  public static boolean isHonorific(String s) {
    return KNOWN_HONORIFICS.contains(s);
  }
}

File: golang/kata/nameinverter/nameinverter.go
package nameinverter

import (
	"fmt"
	"strings"
)

type NameInverter struct {
}

func (n NameInverter) Invert(name string) string {
	nameParts := n.breakingIntoPartsIgnoringWhitespace(name)
	if n.hasOnlyFirstName(nameParts) {
		return nameParts[0] // First name
	}
	return n.invert(n.withoutHonorifics(nameParts))
}

func (n NameInverter) breakingIntoPartsIgnoringWhitespace(name string) []string {
	s := strings.Trim(name, " ")
	names := strings.Split(s, " ")
	return names
}

func (n NameInverter) hasOnlyFirstName(nameParts []string) bool {
	return len(nameParts) < 2
}

func (n NameInverter) withoutHonorifics(nameParts []string) []string {
	if IsHonorific(nameParts[0]) {
		return nameParts[1:]
	}
	return nameParts
}

func (n NameInverter) invert(nameParts []string) string {
	lastName := nameParts[1]
	firstName := nameParts[0]
	postNominal := n.findAndMergerPostnominal(nameParts)

	return strings.Trim(fmt.Sprintf("%s, %s %s", lastName, firstName, postNominal), " ")
}

func (n NameInverter) findAndMergerPostnominal(nameParts []string) string {
	var postNominal string
	for _, s := range nameParts[2:] {
		postNominal += s + " "
	}
	return postNominal
}
